### PR TITLE
Introduce Lombok @Builder and @Value

### PR DESCRIPTION
**Pull request by abdelhamidbakhta**
##  PR description

This PR introduces Lombok `@Builder` and `@Value` annotations. 
- **@Value** : is the immutable variant of `@Data`; all fields are made `private` and `final` by default, and setters are not generated. 
- **@Builder** : The `@Builder` annotation produces complex builder APIs for your classes.. 

A single class has been updated with lombok annotations, as a showcase to illustrate the benefits of using some of the available annotations. The impacted class is : `EthProtocolConfiguration`. It represents a common set of configuration objects used in Besu.

A whitelist mechanism of Lombok annotations will come later. Ultimately we will add rules in `error-prone` to check if only whitelisted annotations are used.

Lines of code reduction: 182L ==> 29L ( =~ **84**% reduction )

### Before

```java
package org.hyperledger.besu.ethereum.eth;

import org.hyperledger.besu.util.number.PositiveNumber;

import java.util.Objects;

import com.google.common.base.MoreObjects;

public class EthProtocolConfiguration {

  public static final int DEFAULT_MAX_GET_BLOCK_HEADERS = 192;
  public static final int DEFAULT_MAX_GET_BLOCK_BODIES = 128;
  public static final int DEFAULT_MAX_GET_RECEIPTS = 256;
  public static final int DEFAULT_MAX_GET_NODE_DATA = 384;
  public static final int DEFAULT_MAX_GET_POOLED_TRANSACTIONS = 256;
  public static final boolean DEFAULT_ETH_65_ENABLED = false;

  private final int maxGetBlockHeaders;
  private final int maxGetBlockBodies;
  private final int maxGetReceipts;
  private final int maxGetNodeData;
  private final int maxGetPooledTransactions;
  private final boolean eth65Enabled;

  public EthProtocolConfiguration(
      final int maxGetBlockHeaders,
      final int maxGetBlockBodies,
      final int maxGetReceipts,
      final int maxGetNodeData,
      final int maxGetPooledTransactions,
      final boolean eth65Enabled) {
    this.maxGetBlockHeaders = maxGetBlockHeaders;
    this.maxGetBlockBodies = maxGetBlockBodies;
    this.maxGetReceipts = maxGetReceipts;
    this.maxGetNodeData = maxGetNodeData;
    this.maxGetPooledTransactions = maxGetPooledTransactions;
    this.eth65Enabled = eth65Enabled;
  }

  public static EthProtocolConfiguration defaultConfig() {
    return new EthProtocolConfiguration(
        DEFAULT_MAX_GET_BLOCK_HEADERS,
        DEFAULT_MAX_GET_BLOCK_BODIES,
        DEFAULT_MAX_GET_RECEIPTS,
        DEFAULT_MAX_GET_NODE_DATA,
        DEFAULT_MAX_GET_POOLED_TRANSACTIONS,
        DEFAULT_ETH_65_ENABLED);
  }

  public static Builder builder() {
    return new Builder();
  }

  public int getMaxGetBlockHeaders() {
    return maxGetBlockHeaders;
  }

  public int getMaxGetBlockBodies() {
    return maxGetBlockBodies;
  }

  public int getMaxGetReceipts() {
    return maxGetReceipts;
  }

  public int getMaxGetNodeData() {
    return maxGetNodeData;
  }

  public int getMaxGetPooledTransactions() {
    return maxGetPooledTransactions;
  }

  public boolean isEth65Enabled() {
    return eth65Enabled;
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    final EthProtocolConfiguration that = (EthProtocolConfiguration) o;
    return maxGetBlockHeaders == that.maxGetBlockHeaders
        && maxGetBlockBodies == that.maxGetBlockBodies
        && maxGetReceipts == that.maxGetReceipts
        && maxGetNodeData == that.maxGetNodeData
        && maxGetPooledTransactions == that.maxGetPooledTransactions;
  }

  @Override
  public int hashCode() {
    return Objects.hash(maxGetBlockHeaders, maxGetBlockBodies, maxGetReceipts, maxGetNodeData);
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("maxGetBlockHeaders", maxGetBlockHeaders)
        .add("maxGetBlockBodies", maxGetBlockBodies)
        .add("maxGetReceipts", maxGetReceipts)
        .add("maxGetNodeData", maxGetNodeData)
        .add("maxGetPooledTransactions", maxGetPooledTransactions)
        .toString();
  }

  public static class Builder {
    private PositiveNumber maxGetBlockHeaders =
        PositiveNumber.fromInt(EthProtocolConfiguration.DEFAULT_MAX_GET_BLOCK_HEADERS);

    private PositiveNumber maxGetBlockBodies =
        PositiveNumber.fromInt(EthProtocolConfiguration.DEFAULT_MAX_GET_BLOCK_BODIES);

    private PositiveNumber maxGetReceipts =
        PositiveNumber.fromInt(EthProtocolConfiguration.DEFAULT_MAX_GET_RECEIPTS);

    private PositiveNumber maxGetNodeData =
        PositiveNumber.fromInt(EthProtocolConfiguration.DEFAULT_MAX_GET_NODE_DATA);

    private PositiveNumber maxGetPooledTransactions =
        PositiveNumber.fromInt(EthProtocolConfiguration.DEFAULT_MAX_GET_POOLED_TRANSACTIONS);

    private boolean eth65Enabled = EthProtocolConfiguration.DEFAULT_ETH_65_ENABLED;

    public Builder maxGetBlockHeaders(final PositiveNumber maxGetBlockHeaders) {
      this.maxGetBlockHeaders = maxGetBlockHeaders;
      return this;
    }

    public Builder maxGetBlockBodies(final PositiveNumber maxGetBlockBodies) {
      this.maxGetBlockBodies = maxGetBlockBodies;
      return this;
    }

    public Builder maxGetReceipts(final PositiveNumber maxGetReceipts) {
      this.maxGetReceipts = maxGetReceipts;
      return this;
    }

    public Builder maxGetNodeData(final PositiveNumber maxGetNodeData) {
      this.maxGetNodeData = maxGetNodeData;
      return this;
    }

    public Builder maxGetPooledTransactions(final PositiveNumber maxGetPooledTransactions) {
      this.maxGetPooledTransactions = maxGetPooledTransactions;
      return this;
    }

    public Builder eth65Enabled(final boolean eth65Enabled) {
      this.eth65Enabled = eth65Enabled;
      return this;
    }

    public EthProtocolConfiguration build() {
      return new EthProtocolConfiguration(
          maxGetBlockHeaders.getValue(),
          maxGetBlockBodies.getValue(),
          maxGetReceipts.getValue(),
          maxGetNodeData.getValue(),
          maxGetPooledTransactions.getValue(),
          eth65Enabled);
    }
  }
}
```
### After
```java
package org.hyperledger.besu.ethereum.eth;

import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class EthProtocolConfiguration {
  @Builder.Default Number maxGetBlockHeaders = 192;
  @Builder.Default Number maxGetBlockBodies = 128;
  @Builder.Default Number maxGetReceipts = 256;
  @Builder.Default Number maxGetNodeData = 384;
  @Builder.Default Number maxGetPooledTransactions = 256;
  @Builder.Default boolean eth65Enabled = false;
}
```
Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

### CODE REVIEW

**Comment by atoulme** [comment]
I don't mind, but let's make sure to add instructions for developers to tune in their IDEs for lombok. That's been the greatest pain point for adoption of this technology in the past.

Also - the logger annotation might save some hassles as well related to configuring the logger.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java [hunk 1]
```diff
@@ -14,169 +14,16 @@
...
-
-import com.google.common.base.MoreObjects;
+import lombok.Builder;
+import lombok.Value;
 
+@Value
+@Builder
 public class EthProtocolConfiguration {
-
-  public static final int DEFAULT_MAX_GET_BLOCK_HEADERS = 192;
```
Comment (shemnon) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java:25]: I think we should keep the static constant variables.  There are many places where we create a builder just to peek the default value.  This should be a true constant.

File: util/src/main/java/org/hyperledger/besu/util/number/PositiveNumber.java [hunk 2]
```diff
@@ -19,14 +19,18 @@
 
 import java.util.Objects;
 
-public class PositiveNumber {
+public class PositiveNumber extends Number {
```
Comment (shemnon) [inline util/src/main/java/org/hyperledger/besu/util/number/PositiveNumber.java:22]: Is this essential to this PR?  Can it be spun out independently and done sooner?

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/EthServerTest.java [hunk 3]
```diff
@@ -55,7 +55,14 @@ public void setUp() {
         worldStateArchive,
         transactionPool,
         ethMessages,
-        new EthProtocolConfiguration(2, 2, 2, 2, 2, true));
+        EthProtocolConfiguration.builder()
+            .maxGetBlockBodies(2)
```
Comment (shemnon) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/EthServerTest.java:59]: Could we make a convenience method `allLimits(int)` that sets these values?  Same with `defaultConfig()` doing `builder().build()`?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java [hunk 4]
```diff
@@ -14,169 +14,16 @@
...
-      return new EthProtocolConfiguration(
-          maxGetBlockHeaders.getValue(),
-          maxGetBlockBodies.getValue(),
-          maxGetReceipts.getValue(),
-          maxGetNodeData.getValue(),
-          maxGetPooledTransactions.getValue(),
-          eth65Enabled);
-    }
-  }
+  @Builder.Default Integer maxGetBlockHeaders = 192;
```
Comment (shemnon) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java:23]: Why the boxed type?  Should it just be `int`?

**Comment by abdelhamidbakhta** [comment]
> I don't mind, but let's make sure to add instructions for developers to tune in their IDEs for lombok. That's been the greatest pain point for adoption of this technology in the past.
> 
> Also - the logger annotation might save some hassles as well related to configuring the logger.

You are totally right. I will bring the topic on rocket chat.

File: util/src/main/java/org/hyperledger/besu/util/number/PositiveNumber.java (hunk 2 above)
Comment (abdelhamidbakhta) [reply util/src/main/java/org/hyperledger/besu/util/number/PositiveNumber.java:22]: It is not essential to this PR

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/EthServerTest.java (hunk 3 above)
Comment (abdelhamidbakhta) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/EthServerTest.java:59]: We could indeed.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java (hunk 4 above)
Comment (abdelhamidbakhta) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java:23]: Yeah i will use the primitive type instead

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java (hunk 1 above)
Comment (abdelhamidbakhta) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java:25]: Done

File: util/src/main/java/org/hyperledger/besu/util/number/PositiveNumber.java (hunk 2 above)
Comment (abdelhamidbakhta) [reply util/src/main/java/org/hyperledger/besu/util/number/PositiveNumber.java]: Removed

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/EthServerTest.java (hunk 3 above)
Comment (abdelhamidbakhta) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/EthServerTest.java]: Done. And actually is it a nice way to show how to write custom setters in Builder

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java (hunk 4 above)
Comment (abdelhamidbakhta) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java]: Done

**Pull request closed without merging**